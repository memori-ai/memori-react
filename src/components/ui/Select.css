.memori-select {
    max-width: 240px;
    min-height: 38px;
}

.memori-select--button {
    position: relative;
    width: 100%;
    max-width: 240px;
    min-height: 38px;
    padding-top: 0.5rem;
    padding-right: 2.5rem;
    padding-bottom: 0.5rem;
    padding-left: 0.75rem;
    border: 1px solid #e5e7eb;
    border-radius: 0.5rem;
    background: transparent;
    box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(0, 0, 0, 0.1) 0px 4px 6px -1px,
        rgba(0, 0, 0, 0.1) 0px 2px 4px -2px;
    color: var(--memori-text-color, #333);
    cursor: pointer;
    font-family: var(--memori-font-family);
    font-size: 16px;
    text-align: left;
}

.memori-select--button:focus {
    border-color: #cbd5e0;
    box-shadow: 0 0 0 3px rgba(66, 153, 225, 0.5);
}

.memori-select--button:focus-visible {
    border-color: #63b3ed;
    box-shadow: 0 0 0 3px rgba(66, 153, 225, 0.5);
}

.memori-select--button:hover {
    border-color: #e2e8f0;
}

.memori-select--button:active {
    border-color: #e2e8f0;
}

.memori-select--button:disabled {
    border-color: #e2e8f0;
    background-color: #f7fafc;
    color: #a0aec0;
    cursor: not-allowed;
}

@media (min-width: 640px) {
    .memori-select--button {
        font-size: 0.875rem;
        line-height: 1.25rem;
    }
}

.memori-select--label {
    display: inline-block;
    margin-bottom: 0.25rem;
}

.memori-select--value {
    display: block;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
}

.memori-select--value.memori-select--value-placeholder {
    color: #a0aec0;
}

.memori-select--icon {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    display: flex;
    align-items: center;
    padding-right: 0.5rem;
    pointer-events: none;
}

.memori-select--icon svg {
    width: 1.25rem;
    height: 1.25rem;
    color: rgb(156 163 175/1);
}

ul.memori-select--options {
    position: absolute;
    z-index: 1;
    overflow: auto;
    width: 100%;
    max-width: min(18rem, 30%);
    max-height: 15rem;
    padding-top: 0.25rem;
    padding-right: 0;
    padding-bottom: 0.25rem;
    padding-left: 0;
    border-radius: 0.375rem;
    margin-top: 0.25rem;
    margin-right: 0;
    margin-bottom: 0;
    margin-left: 0;
    background: #fff;
    box-shadow: rgb(255, 255, 255) 0px 0px 0px 0px, rgba(0, 0, 0, 0.05) 0px 0px 0px 1px,
        rgba(0, 0, 0, 0.1) 0px 10px 15px -3px, rgba(0, 0, 0, 0.1) 0px 4px 6px -4px;
    list-style: none;
}

@media (min-width: 640px) {
    .memori-select--options {
        font-size: 0.875rem;
        line-height: 1.25rem;
    }
}

li.memori-select--option {
    position: relative;
    padding-top: 0.5rem;
    padding-right: 1rem;
    padding-bottom: 0.5rem;
    padding-left: 1rem;
    color: rgb(17 24 39/1);
    cursor: pointer;
    user-select: none;
}

li.memori-select--option:hover,
li.memori-select--option:focus {
    background-color: #f7fafc;
}

li.memori-select--option:focus {
    outline: 2px solid #63b3ed;
    outline-offset: 2px;
}