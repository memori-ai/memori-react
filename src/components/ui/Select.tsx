import React, { Fragment } from 'react';
import cx from 'classnames';
import { Listbox, Transition } from '@headlessui/react';
import SelectIcon from '../icons/SelectIcon';

export interface Props<T = any> {
  className?: string;
  value?: T;
  name?: string;
  displayValue?: string;
  onChange: (value: T) => void;
  options: { value: T; label: any }[];
  disabled?: boolean;
  label?: string;
  placeholder?: string;
}

const Select = ({
  className,
  value,
  name,
  displayValue,
  options,
  onChange,
  disabled = false,
  label,
  placeholder,
}: Props) => {
  return (
    <div className={cx('memori-select', className)}>
      <Listbox
        value={value}
        onChange={value => {
          onChange(value);
        }}
        disabled={disabled}
        name={name}
      >
        {label && (
          <Listbox.Label className="memori-select--label">
            {label}:
          </Listbox.Label>
        )}
        <Listbox.Button aria-label={label} className="memori-select--button">
          <span
            className={cx('memori-select--value', {
              'memori-select--value-placeholder': !value,
            })}
          >
            {value ? displayValue || value : placeholder}
          </span>
          <span className="memori-select--icon">
            <SelectIcon />
          </span>
        </Listbox.Button>
        <Transition
          as={Fragment}
          leave="transition ease-in duration-100"
          leaveFrom="opacity-100"
          leaveTo="opacity-0"
        >
          <Listbox.Options className="memori-select--options">
            {options.map(opt => (
              <Listbox.Option
                key={opt.value}
                value={opt.value}
                className="memori-select--option"
              >
                {opt.label}
              </Listbox.Option>
            ))}
          </Listbox.Options>
        </Transition>
      </Listbox>
    </div>
  );
};

export default Select;
