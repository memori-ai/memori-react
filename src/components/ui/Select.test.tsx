import React from 'react';
import { render } from '@testing-library/react';
import Select from './Select';

it('renders Select unchanged', () => {
  const { container } = render(
    <Select
      label="Select me"
      options={[
        { value: '1', label: '1' },
        { value: '2', label: '2' },
        { value: '3', label: '3' },
      ]}
      onChange={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders Select with placeholder unchanged', () => {
  const { container } = render(
    <Select
      label="Select me"
      placeholder="Select me"
      options={[
        { value: '1', label: '1' },
        { value: '2', label: '2' },
        { value: '3', label: '3' },
      ]}
      onChange={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders Select with value unchanged', () => {
  const { container } = render(
    <Select
      label="Select me"
      value="2"
      options={[
        { value: '1', label: '1' },
        { value: '2', label: '2' },
        { value: '3', label: '3' },
      ]}
      onChange={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders Select with custom value display unchanged', () => {
  const { container } = render(
    <Select
      label="Select me"
      value="2"
      displayValue="Two"
      options={[
        { value: '1', label: 'One' },
        { value: '2', label: 'Two' },
        { value: '3', label: 'Three' },
      ]}
      onChange={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

it('renders Select disabled unchanged', () => {
  const { container } = render(
    <Select
      label="Select me"
      value="2"
      options={[
        { value: '1', label: '1' },
        { value: '2', label: '2' },
        { value: '3', label: '3' },
      ]}
      disabled
      onChange={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});
