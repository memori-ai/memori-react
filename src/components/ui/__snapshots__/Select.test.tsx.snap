// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Select disabled unchanged 1`] = `
<div>
  <div
    class="memori-select"
  >
    <label
      class="memori-select--label"
      data-headlessui-state="disabled"
      id="headlessui-listbox-label-:r8:"
    >
      Select me
      :
    </label>
    <button
      aria-haspopup="true"
      aria-label="Select me"
      aria-labelledby="headlessui-listbox-label-:r8: headlessui-listbox-button-:r9:"
      class="memori-select--button"
      data-headlessui-state="disabled"
      disabled=""
      id="headlessui-listbox-button-:r9:"
      type="button"
    >
      <span
        class="memori-select--value"
      >
        2
      </span>
      <span
        class="memori-select--icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          role="img"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 3a.75.75 0 01.55.24l3.25 3.5a.75.75 0 11-1.1 1.02L10 4.852 7.3 7.76a.75.75 0 01-1.1-1.02l3.25-3.5A.75.75 0 0110 3zm-3.76 9.2a.75.75 0 011.06.04l2.7 2.908 2.7-2.908a.75.75 0 111.1 1.02l-3.25 3.5a.75.75 0 01-1.1 0l-3.25-3.5a.75.75 0 01.04-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`renders Select unchanged 1`] = `
<div>
  <div
    class="memori-select"
  >
    <label
      class="memori-select--label"
      data-headlessui-state=""
      id="headlessui-listbox-label-:r0:"
    >
      Select me
      :
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Select me"
      aria-labelledby="headlessui-listbox-label-:r0: headlessui-listbox-button-:r1:"
      class="memori-select--button"
      data-headlessui-state=""
      id="headlessui-listbox-button-:r1:"
      type="button"
    >
      <span
        class="memori-select--value memori-select--value-placeholder"
      />
      <span
        class="memori-select--icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          role="img"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 3a.75.75 0 01.55.24l3.25 3.5a.75.75 0 11-1.1 1.02L10 4.852 7.3 7.76a.75.75 0 01-1.1-1.02l3.25-3.5A.75.75 0 0110 3zm-3.76 9.2a.75.75 0 011.06.04l2.7 2.908 2.7-2.908a.75.75 0 111.1 1.02l-3.25 3.5a.75.75 0 01-1.1 0l-3.25-3.5a.75.75 0 01.04-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`renders Select with custom value display unchanged 1`] = `
<div>
  <div
    class="memori-select"
  >
    <label
      class="memori-select--label"
      data-headlessui-state=""
      id="headlessui-listbox-label-:r6:"
    >
      Select me
      :
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Select me"
      aria-labelledby="headlessui-listbox-label-:r6: headlessui-listbox-button-:r7:"
      class="memori-select--button"
      data-headlessui-state=""
      id="headlessui-listbox-button-:r7:"
      type="button"
    >
      <span
        class="memori-select--value"
      >
        Two
      </span>
      <span
        class="memori-select--icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          role="img"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 3a.75.75 0 01.55.24l3.25 3.5a.75.75 0 11-1.1 1.02L10 4.852 7.3 7.76a.75.75 0 01-1.1-1.02l3.25-3.5A.75.75 0 0110 3zm-3.76 9.2a.75.75 0 011.06.04l2.7 2.908 2.7-2.908a.75.75 0 111.1 1.02l-3.25 3.5a.75.75 0 01-1.1 0l-3.25-3.5a.75.75 0 01.04-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`renders Select with placeholder unchanged 1`] = `
<div>
  <div
    class="memori-select"
  >
    <label
      class="memori-select--label"
      data-headlessui-state=""
      id="headlessui-listbox-label-:r2:"
    >
      Select me
      :
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Select me"
      aria-labelledby="headlessui-listbox-label-:r2: headlessui-listbox-button-:r3:"
      class="memori-select--button"
      data-headlessui-state=""
      id="headlessui-listbox-button-:r3:"
      type="button"
    >
      <span
        class="memori-select--value memori-select--value-placeholder"
      >
        Select me
      </span>
      <span
        class="memori-select--icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          role="img"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 3a.75.75 0 01.55.24l3.25 3.5a.75.75 0 11-1.1 1.02L10 4.852 7.3 7.76a.75.75 0 01-1.1-1.02l3.25-3.5A.75.75 0 0110 3zm-3.76 9.2a.75.75 0 011.06.04l2.7 2.908 2.7-2.908a.75.75 0 111.1 1.02l-3.25 3.5a.75.75 0 01-1.1 0l-3.25-3.5a.75.75 0 01.04-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`renders Select with value unchanged 1`] = `
<div>
  <div
    class="memori-select"
  >
    <label
      class="memori-select--label"
      data-headlessui-state=""
      id="headlessui-listbox-label-:r4:"
    >
      Select me
      :
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Select me"
      aria-labelledby="headlessui-listbox-label-:r4: headlessui-listbox-button-:r5:"
      class="memori-select--button"
      data-headlessui-state=""
      id="headlessui-listbox-button-:r5:"
      type="button"
    >
      <span
        class="memori-select--value"
      >
        2
      </span>
      <span
        class="memori-select--icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          focusable="false"
          role="img"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 3a.75.75 0 01.55.24l3.25 3.5a.75.75 0 11-1.1 1.02L10 4.852 7.3 7.76a.75.75 0 01-1.1-1.02l3.25-3.5A.75.75 0 0110 3zm-3.76 9.2a.75.75 0 011.06.04l2.7 2.908 2.7-2.908a.75.75 0 111.1 1.02l-3.25 3.5a.75.75 0 01-1.1 0l-3.25-3.5a.75.75 0 01.04-1.06z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
